### PR TITLE
ARROW-7691: [C++] Check non-scalar Flatbuffers fields are not null

### PR DESCRIPTION
**Pull request by pitrou**
We're discussing whether to make those fields required in the schema definitions (which would make validation automatic by the flatbuffers generated verifier), but in the meantime we can check those fields manually.

This should fix a bunch of issues detected by OSS-Fuzz.

### CODE REVIEW

**Comment by pitrou** [comment]
cc @emkornfield, @fsaintjacques

File: cpp/src/arrow/ipc/metadata_internal.h [hunk 1]
```diff
@@ -92,6 +88,21 @@ struct FileBlock {
...
 
+// Low-level utilities to help with reading Flatbuffers data.
+
+#define CHECK_FLATBUFFERS_NOT_NULL(fb_value, name)             \
+  if ((fb_value) == NULLPTR) {                                 \
+    return Status::IOError("Unexpected null field ", name,     \
+                           " in flatbuffer-encoded metadata"); \
+  }
+
+#define FLATBUFFERS_VECTOR_SIZE(fb_vector) \
```
Comment (emkornfield) [inline cpp/src/arrow/ipc/metadata_internal.h]: nit, does this need to be a macro?

**Comment by emkornfield** [comment]
LGTM, 1 minor comment.

File: cpp/src/arrow/ipc/metadata_internal.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/ipc/metadata_internal.h]: Probably not indeed.

**Pull request closed without merging**